### ISSUE TITLE
gRPC Query erroring with dictionary usize issue

### ISSUE THREAD

**Issue by e-dard** [influxdata/influxdb_iox#2112]
The following Flux is causing IOx to error in our internal deployment:

```
from(bucket: "apps")
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_measurement"] == "http_query_response_bytes")
  |> aggregateWindow(every: v.windowPeriod, fn: count, createEmpty: false)
```

This results in the `read_window_aggregate` physical operation being issues, which then needs a `Grouped SeriesSet` plan to execute in order to get the data in the correct series-key based format.

The query never works and Flux displays `An internal error has occurred`.

I can see the following in the IOx logs:

```
level=info msg=read_window_aggregate 
db_name=844910ece80be8bc_3c0bd4c89186ca89 
range="Some(TimestampRange { start: 1627303947974188978, end: 1627304247974188978 })" 
window_every=0 
offset=0 
aggregate="[Aggregate { r#type: Count }]" 
window="Some(Window { every: Some(Duration { nsecs: 10000000000, months: 0, negative: false }), offset: Some(Duration { nsecs: 0, months: 0, negative: false }) })" 
predicate="(TagRef:_m[0x00] == \"http_query_response_bytes\")" target="influxdb_iox::influxdb_ioxd::rpc::storage::service" location="src/influxdb_ioxd/rpc/storage/service.rs:357" time=1627304247996674889
```

There is also the following error:

```
level=error msg="Error Running Grouped SeriesSet 
Plan: GroupingSeries { 
    db_name: \"844910ece80be8bc_3c0bd4c89186ca89\", 
    source: SeriesSetConversion { 
        source: ReadingRecordBatch { 
            source: ExternalError(
               ArrowError(ExternalError(
                     ArrowError(ExternalError(Internal(\"Can not convert index to usize in dictionary of type creating group by value 
                      Int32\")))))) } } }" 
          target="data_types::error" location="data_types/src/error.rs:18" time=1627304267133351961
```

I believe that this is a reproducer for https://github.com/influxdata/influxdb_iox/issues/1515....

**Comment by alamb** [comment]
I have narrowed the problem down to a single column in the `http_query_response_bytes` table: `org_id`

I will now attempt to figure out what is going on with that column

```shell
(arrow_dev) alamb@MacBook-Pro:~/Software/influxdb_iox$   ./target/debug/influxdb_iox  --host http://localhost:1234 database query 844910ece80be8bc_3c0bd4c89186ca89 "select count(distinct org_id) from http_query_response_bytes;"
Error querying: status: Internal, message: "Internal error reading points from database 844910ece80be8bc_3c0bd4c89186ca89:  Internal error executing plan: Arrow error: External error: Internal error: Can not convert index to usize in dictionary of type creating group by value Int32. This was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker", details: [], metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Mon, 26 Jul 2021 17:20:15 GMT"} }
```

**Comment by alamb** [comment]
I have dug another panic out of the log related to bitmap construction.

I think I was careful to try and avoid any data that was in the MUB with this query (and thus maybe this means bad data has gotten into the RUB?) but I am not sure and that may not be correct. 

```


thread 'IOx Query Executor Thread' panicked at 'assertion failed: i < (self.bits.len() << 3)', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/arrow-5.0.0/src/bitmap.rs:56:9
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
level=error msg="thread 'IOx Query Executor Thread' panicked at 'assertion failed: i < (self.bits.len() << 3)', /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/arrow-5.0.0/src/bitmap.rs:56:9" target="panic_logging" location="panic_logging/src/lib.rs:110" time=1627324191388471358
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/panicking.rs:50:5
   3: arrow::bitmap::Bitmap::is_set
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/arrow-5.0.0/src/bitmap.rs:56:9
   4: arrow::array::data::ArrayData::is_null
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/arrow-5.0.0/src/array/data.rs:292:21
   5: arrow::array::array::Array::is_null
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/arrow-5.0.0/src/array/array.rs:164:9
   6: datafusion::scalar::ScalarValue::try_from_array
             at ./usr/local/cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/30693df/datafusion/src/scalar.rs:687:31
   7: datafusion::scalar::ScalarValue::try_from_array
   8: datafusion::physical_plan::Accumulator::update_batch::{{closure}}::{{closure}}
             at ./usr/local/cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/30693df/datafusion/src/physical_plan/mod.rs:572:30
   9: core::iter::adapters::map::map_try_fold::{{closure}}
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/adapters/map.rs:89:28
  10: core::iter::traits::iterator::Iterator::try_fold
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/traits/iterator.rs:2006:21
  11: <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::try_fold
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/adapters/map.rs:115:9
  12: <core::iter::adapters::ResultShunt<I,E> as core::iter::traits::iterator::Iterator>::try_fold
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/adapters/mod.rs:173:9
  13: core::iter::traits::iterator::Iterator::find
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/traits/iterator.rs:2358:9
  14: <core::iter::adapters::ResultShunt<I,E> as core::iter::traits::iterator::Iterator>::next
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/adapters/mod.rs:155:9
  15: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter_nested::SpecFromIterNested<T,I>>::from_iter
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/alloc/src/vec/spec_from_iter_nested.rs:23:32
  16: <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/alloc/src/vec/spec_from_iter.rs:36:9
  17: <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/alloc/src/vec/mod.rs:2404:9
  18: core::iter::traits::iterator::Iterator::collect
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/traits/iterator.rs:1788:9
  19: <core::result::Result<V,E> as core::iter::traits::collect::FromIterator<core::result::Result<A,E>>>::from_iter::{{closure}}
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/result.rs:1625:53
  20: core::iter::adapters::process_results
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/adapters/mod.rs:144:17
  21: <core::result::Result<V,E> as core::iter::traits::collect::FromIterator<core::result::Result<A,E>>>::from_iter
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/result.rs:1625:9
  22: core::iter::traits::iterator::Iterator::collect
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/traits/iterator.rs:1788:9
  23: datafusion::physical_plan::Accumulator::update_batch::{{closure}}
             at ./usr/local/cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/30693df/datafusion/src/physical_plan/mod.rs:570:21
  24: core::iter::traits::iterator::Iterator::try_for_each::call::{{closure}}
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/traits/iterator.rs:2048:26
  25: core::iter::traits::iterator::Iterator::try_fold
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/traits/iterator.rs:2006:21
  26: core::iter::traits::iterator::Iterator::try_for_each
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/traits/iterator.rs:2051:9
  27: datafusion::physical_plan::Accumulator::update_batch
             at ./usr/local/cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/30693df/datafusion/src/physical_plan/mod.rs:569:9
  28: datafusion::physical_plan::hash_aggregate::aggregate_batch::{{closure}}
             at ./usr/local/cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/30693df/datafusion/src/physical_plan/hash_aggregate.rs:931:43
  29: core::iter::traits::iterator::Iterator::try_for_each::call::{{closure}}
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/traits/iterator.rs:2048:26
  30: core::iter::traits::iterator::Iterator::try_fold
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/traits/iterator.rs:2006:21
  31: core::iter::traits::iterator::Iterator::try_for_each
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/iter/traits/iterator.rs:2051:9
  32: datafusion::physical_plan::hash_aggregate::aggregate_batch
             at ./usr/local/cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/30693df/datafusion/src/physical_plan/hash_aggregate.rs:918:5
  33: datafusion::physical_plan::hash_aggregate::compute_hash_aggregate::{{closure}}
             at ./usr/local/cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/30693df/datafusion/src/physical_plan/hash_aggregate.rs:872:9
  34: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/future/mod.rs:80:19
  35: datafusion::physical_plan::hash_aggregate::HashAggregateStream::new::{{closure}}
             at ./usr/local/cargo/git/checkouts/arrow-datafusion-71ae82d9dec9a01c/30693df/datafusion/src/physical_plan/hash_aggregate.rs:895:17
  36: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/future/mod.rs:80:19
  37: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/core.rs:243:17
  38: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/loom/std/unsafe_cell.rs:14:9
  39: tokio::runtime::task::core::CoreStage<T>::poll
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/core.rs:233:13
  40: tokio::runtime::task::harness::poll_future::{{closure}}
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/harness.rs:438:23
  41: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panic.rs:344:9
  42: std::panicking::try::do_call
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:379:40
  43: std::panicking::try
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:343:19
  44: std::panic::catch_unwind
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panic.rs:431:14
  45: tokio::runtime::task::harness::poll_future
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/harness.rs:425:19
  46: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/harness.rs:89:9
  47: tokio::runtime::task::harness::Harness<T,S>::poll
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/harness.rs:59:15
  48: tokio::runtime::task::raw::RawTask::poll
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/raw.rs:70:18
  49: tokio::runtime::task::Notified<S>::run
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/mod.rs:171:9
  50: tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/thread_pool/worker.rs:370:21
  51: tokio::coop::with_budget::{{closure}}
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/coop.rs:106:9
  52: std::thread::local::LocalKey<T>::try_with
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/thread/local.rs:376:16
  53: std::thread::local::LocalKey<T>::with
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/thread/local.rs:352:9
  54: tokio::coop::with_budget
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/coop.rs:99:5
  55: tokio::coop::budget
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/coop.rs:76:5
  56: tokio::runtime::thread_pool::worker::Context::run_task
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/thread_pool/worker.rs:348:9
  57: tokio::runtime::thread_pool::worker::Context::run
  58: tokio::runtime::thread_pool::worker::run::{{closure}}
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/thread_pool/worker.rs:303:17
  59: tokio::macros::scoped_tls::ScopedKey<T>::set
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/macros/scoped_tls.rs:61:9
  60: tokio::runtime::thread_pool::worker::run
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/thread_pool/worker.rs:300:5
  61: tokio::runtime::thread_pool::worker::Launch::launch::{{closure}}
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/thread_pool/worker.rs:279:45
  62: <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/blocking/task.rs:42:21
  63: tokio::runtime::task::core::CoreStage<T>::poll::{{closure}}
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/core.rs:243:17
  64: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/loom/std/unsafe_cell.rs:14:9
  65: tokio::runtime::task::core::CoreStage<T>::poll
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/core.rs:233:13
  66: tokio::runtime::task::harness::poll_future::{{closure}}
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/harness.rs:438:23
  67: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panic.rs:344:9
  68: std::panicking::try::do_call
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:379:40
  69: std::panicking::try
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:343:19
  70: std::panic::catch_unwind
             at ./rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panic.rs:431:14
  71: tokio::runtime::task::harness::poll_future
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/harness.rs:425:19
  72: tokio::runtime::task::harness::Harness<T,S>::poll_inner
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/harness.rs:89:9
  73: tokio::runtime::task::harness::Harness<T,S>::poll
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/harness.rs:59:15
  74: tokio::runtime::task::raw::RawTask::poll
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/raw.rs:70:18
  75: tokio::runtime::task::Notified<S>::run
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/task/mod.rs:171:9
  76: tokio::runtime::blocking::pool::Inner::run
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/blocking/pool.rs:265:17
  77: tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}
             at ./usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.2/src/runtime/blocking/pool.rs:245:17
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by alamb** [comment]
Woohoo ! I have a (local) reproducer 🕵️  

It happens if you write this data into the MUB (in three distinct writes, not sure if that is important yet)
```
http_query_response_bytes,tag1=first1,tag2=second1 counter=0 1627329202131277000
```
Then
```
http_query_response_bytes,tag1=first1,tag2=second1 counter=0 1627329202143546000
```
Then
```
http_query_response_bytes,tag1=first1,tag2=second1 counter=0 1627329202152517000
http_query_response_bytes,tag2=second2 counter=0 1627329202152517001
```


The full on repro looks like


```shell
cargo run -- run   --object-store=file --data-dir=$HOME/.influxdb_iox --server-id=42
./target/debug/influxdb_iox database create test_db
```

And the you can run this rust script (also attached): [repro.zip](https://github.com/influxdata/influxdb_iox/files/6880857/repro.zip)


```rust

use influxdb_iox_client::write::Client as WriteClient;
use influxdb_iox_client::flight::Client as FlightClient;
use std::io::Write;

const GRPC_BASE: &str = "http://localhost:8082";
const DB_NAME: &str = "test_db";

#[tokio::main]
async fn main() {
    println!("Connecting...");
    let mut w = write_client().await;
    let mut f = flight_client().await;

    write_some_data(&data2(0), &mut w, &mut f).await;
    write_some_data(&data2(1), &mut w, &mut f).await;
    write_some_data(&data2(2), &mut w, &mut f).await;



    // for num_rows in 0..100 {
    //     println!("Writing num_rows: {}", num_rows);
    //     write_some_data(&data2(num_rows), &mut w, &mut f).await;
    // }
}

// This returns data in the following pattern:
//
// tag1:Value, Tag2:Value
//
// tag1:Null, Tag2:Value
// tag1:Value, Tag2:Value
//
// tag1:Null, Tag2:Value
// tag1:Null, Tag2:Value
// tag1:Value, Tag2:Value
//
// ...
fn data2(desired_rows: usize) -> String {
    let mut data: Vec<u8> = vec![];
    let mut now_ts = chrono::Utc::now().timestamp_nanos();

    let mut tag1_offset = 0;
    let mut num_rows = 0;
    'outer: loop {
        let tag1_value = format!("first{}", tag1_offset);
        let tag2_value = format!("second{}", tag1_offset);
        for row in 0..tag1_offset {
            // last row
            if row + 1 == tag1_offset {
                writeln!(data, "http_query_response_bytes,tag1={},tag2={} counter={} {}", tag1_value, tag2_value, row, now_ts).unwrap();
            }
            else {
                writeln!(data, "http_query_response_bytes,tag2={} counter={} {}", tag2_value, row, now_ts).unwrap();
            }
            num_rows += 1;
            now_ts += 1;
            if num_rows >= desired_rows {
                break 'outer
            }
        }

        tag1_offset += 1;
    }
    String::from_utf8(data).expect("valid utf8")
}





/// The basic idea here is to write data and then query IOx to ensure we don't hit any error
async fn write_some_data(lp: &str, w: &mut WriteClient, f: &mut FlightClient) {
    println!("Writing {} bytes o data in ur database", lp.len());
    println!("{}", lp);

    match w.write(DB_NAME, lp).await {
        Ok(_) => {}
        Err(e) => {
            panic!("Error writing data: {}\n{:?}\n\n{}", e, e, lp)
        }
    };

    // now, do a query and validate that still works
    let sql_query = "select count(distinct tag1), count(distinct tag2) from http_query_response_bytes";

    let mut query_results = f
        .perform_query(DB_NAME, sql_query)
        .await
        .unwrap();


    loop {
        match query_results.next().await {
            Ok(None) => {
                println!("Done receiving query results");
                break
            }
            Ok(Some(data)) => {
                println!("received batch: {} rows", data.num_rows())
            }
            Err(e) => {
                panic!("Error running '{}': {}\n{:?}", sql_query, e, e)
            }
        }
    }
}




/// Return a write client suitable for communicating with this
/// server
async fn write_client() -> WriteClient {
    WriteClient::new(grpc_channel().await)
}

/// Return a flight client suitable for communicating with this
/// server
async fn flight_client() -> FlightClient {
    FlightClient::new(grpc_channel().await)
}


async fn grpc_channel() -> tonic::transport::Channel {
    match tonic::transport::Channel::from_static(GRPC_BASE)
        .connect()
        .await
    {
        Ok(channel) => channel,
        Err(e) => {
            panic!("Error connecting grpc channel to {}: {}", GRPC_BASE, e)
        }
    }
}
```

**Comment by alamb** [comment]
The data is entirely in the mutable buffer:
```
test_db> select * from system.chunks;
+----+---------------------+---------------------------+-------------------+------------------+--------------+--------------------+-----------+--------------------------------+-----------------------------+-----------------------------+-------------+
| id | partition_key       | table_name                | storage           | lifecycle_action | memory_bytes | object_store_bytes | row_count | time_of_last_access            | time_of_first_write         | time_of_last_write          | time_closed |
+----+---------------------+---------------------------+-------------------+------------------+--------------+--------------------+-----------+--------------------------------+-----------------------------+-----------------------------+-------------+
| 0  | 2021-07-26 20:00:00 | http_query_response_bytes | OpenMutableBuffer |                  | 228          |                    | 4         | 2021-07-26T20:14:44.490897975Z | 2021-07-26T20:13:01.287891Z | 2021-07-26T20:13:02.124801Z |             |
+----+---------------------+---------------------------+-------------------+------------------+--------------+--------------------+-----------+--------------------------------+-----------------------------+-----------------------------+-------------+

Returned 1 row in 125.721375ms
```

**Comment by alamb** [comment]
🤔  I wonder if this is a bug in datafusion with not handling nulls in dictionaries correctly. I will see if I can get a reproducer

**Comment by alamb** [comment]
So I am fairly confident that this is a bug in DataFusion handling code for Dictionaries (in code that I contributed 😞 ) -- basically whenaggregatingg, DataFusion is not checking of the validity of the row before. I am working on a DF reproducer that shows this behavior

**Comment by alamb** [comment]
Here is a smaller reproducer (add to `query_tests/src/sql.rs`)

```rust

#[tokio::test]
async fn sql_distinct_aggregates() {
    // validate distinct aggregates work against dictionary columns
    // which have nulls in them
    let expected = vec![
        "| TODO |",
        "+--------+",
    ];
    run_sql_test_case!(
        TwoMeasurementsManyNulls {},
        "select count(distinct city) from o2",
        &expected
    );
}
```

Which results in the symptom we have been seeing:

---- sql::sql_distinct_aggregates stdout ----
Running scenario 'Data in single open chunk of mutable buffer'
SQL: '"select count(distinct city) from o2"'
thread 'sql::sql_distinct_aggregates' panicked at 'Running plan: DataFusionExecution { source: ArrowError(ExternalError(Internal("Can not convert index to usize in dictionary of type creating group by value Int32"))) }', query_tests/src/sql.rs:450:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by alamb** [comment]
Filed root cause ticket in DataFusion: https://github.com/apache/arrow-datafusion/issues/781

**Comment by alamb** [comment]
1. DataFusion GROUP BY doesn’t handle NULLs  correctly for any type :sadparrot: -- https://github.com/apache/arrow-datafusion/issues/781 --  but most of the time it “just gets the wrong answer” rather than panicing
2. Due to how we constructed a DictionaryArray we happened to get a panic  rather than wrong answers but the root cause is the same 
3. I am working on a fix… but it is going to take me longer than I expected (though will have some nice benefits to the rest of the DF community — namely correct answers!)
